wgpu-hal/src/metal/surface.rs

```diff
@@ -186,9 +186,21 @@ impl crate::Surface<super::Api> for super::Surface {
         let drawable_size = CGSize::new(config.extent.width as f64, config.extent.height as f64);
 
         match config.composite_alpha_mode {
-            crate::CompositeAlphaMode::Opaque => render_layer.set_opaque(true),
-            crate::CompositeAlphaMode::PostMultiplied => render_layer.set_opaque(false),
-            crate::CompositeAlphaMode::PreMultiplied => (),
+            wgt::CompositeAlphaMode::Opaque => render_layer.set_opaque(true),
```

Comment:
Does this mean that when composite_alpha_mode is not Opaque, the background will always be clear color? Even when the background color used to be some other color with non opaque alpha?

Reply:
The metalLayer's background color is like the color of a canvas itself, while ColorAttachment's wgpu::LoadOp::Clear is the color to fill the specified area of the canvas. 
```
metal_layer.set_opaque(false); 
msg_send![metal_layer, setBackgroundColor: clear_color];
```
these code means that the canvas itself is transparent, and any colors drawn on this canvas with transparency will be composited by the layer behind it, which is exactly the purpose of using non-opaque AlphaMode.

Reply:
I understand, but I'm only wondering if it is ok to override the user-set background color (which could or could not be opaque). Would the user expect that when he sets the background to (1.0, 0.0, 0.0, 0.5) that (s)he gets a half transparent red background?
I thought the CALayer's isOpaque flag indicates if the opacity component of the content when drawing into it should be ignored or not (for layer composition efficiency reasons) -> https://developer.apple.com/documentation/quartzcore/calayer/1410763-isopaque
It doesn't mention that the background color should be fully transparent (clear color).
We currently use a fork of wgpu and patch ``` crate::CompositeAlphaMode::Opaque => render_layer.set_opaque(true) ``` to call ```set_opaque(false)``` instead. Without any changes to the metal layer's background, we are getting the result we need (we never explicitly set the MetalLayer's background anywhere). So I'm assuming you were setting the layer's background somewhere else (can't find where wgpu's metal backend would be changing it)?

Reply:
As my code comment mentions the reason：
```
// CALayer may have a background color by default,
// if the alpha_mode of the surface is set to non-opaque, unexpected colors are obtained.
```
It's also possible to remove `setBackgroundColor`  line of code, which is only about how we think about: the `metal_layer` may have been created by a window lib and not directly accessible to the user.

Reply:
With unexpected colors, you mean incorrect colors? Or do you mean that because the background color is not nil (default for a MetalLayer) and was set (by the application or external window lib) to a not 100% transparent color, that you get blended colors with that background from the MetalLayer (because it's not clear)?

Reply:
I mean the second case

Reply:
Isn't it then up to that window lib to allow for controller the CAMetalLayer's backgroundColor property instead of overriding the one set by users here in wgpu? Somewhere you are passing your UIView to wpgu, I assume you can use that integration point to alter the view's main layer's backgroundColor.
We are working on an iOS app, so in our case wpgu always will create the CAMetalLayer, as the view's main layer can't be changed on iOS, and that way the metal_layer's background is always nil.
I'd prefer wpgu not to alter background colors of layer implicitely without a way to opt-in/out.

Reply:
Thanks @jinleili for the changes!